### PR TITLE
Yangzhang/codonfm 10b

### PR DESCRIPTION
**Pull request by yzhang123**
### Description

add fsdp2 support for TE

#### Usage

```python
python -m src.runner pretrain \
    --exp_name "$exp_name" \
    --model_name encodon_80m \
    --data_path /data/ncbi/processed_unfiltered/ \
    --process_item mlm_memmap \
    --dataset_name CodonMemmapDataset \
    --lr $learning_rate \
    --num_gpus $num_gpus \
    --num_nodes $num_nodes \
    --train_batch_size $train_batch_size \
    --val_batch_size $val_batch_size \
    --num_workers $num_workers \
    --collate_fn thd \
    --attn_input_format thd \
    --use_transformer_engine \
    --bf16 \
    --split_name_prefix nopathogen \
    --use_transformer_engine \
    --checkpoints_dir results/${exp_name}/checkpoints/ \
    --out_dir results/${exp_name}/ \
   --enable_fsdp

```

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [x] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [x] I have added/updated tests as needed
- [x] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/recipes/codonfm_ptl_te/src/models/components/encodon_te.py [hunk 1]
```diff
@@ -91,25 +92,18 @@ def __init__(self, config):
         )
 
         self.cls = nn.Sequential(
-            transformer_engine.pytorch.Linear(config.hidden_size, config.hidden_size),
+            transformer_engine.pytorch.Linear(config.hidden_size, config.hidden_size, device="cpu"),
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/codonfm_ptl_te/src/models/components/encodon_te.py:95]: why does this move to cpu?

File: bionemo-recipes/recipes/codonfm_ptl_te/src/utils/fsdp_config.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            cpu_offload=cpu_offload,
+            # FSDP2-specific optimizations
+            sharding_strategy="FULL_SHARD",  # Each GPU holds 1/N of parameters
+            state_dict_type="full",
+            # FSDP2 improvements that reduce memory usage
+            use_orig_params=True,  # Better memory management and optimizer compatibility
+            limit_all_gathers=True,  # Reduce memory spikes during forward pass
+        )
+    else:
+        print("Using FSDP1 (consider upgrading to PyTorch 2.1+ for FSDP2 benefits)")
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/codonfm_ptl_te/src/utils/fsdp_config.py]: nit: use logger instead?

File: bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+import torch
+
+
+requires_multi_gpu = pytest.mark.skipif(
+    not torch.cuda.is_available() or torch.cuda.device_count() < 2,
+    reason="Test requires at least 2 GPUs",
+)
+
+
+def run_distributed_training_script(script_path: Path, num_gpus: int = 2, timeout: int = 120):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py]: I see you've made this function but you only call it once. So it might be simpler to put it inline.

File: bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+@requires_multi_gpu
+def test_encodon_pl_fsdp_training_2gpus(tmp_path):
+    """Test EncodonPL training with FSDP on 2 GPUs.
+
+    This test runs as a subprocess with torchrun because FSDP requires
+    proper distributed process initialization that cannot be done within pytest.
+    """
+
+    # Create a self-contained training script
+    training_script = textwrap.dedent("""
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py]: I'm a bit concerned about this test since it will break the moment you change the dataset / training script since it's not testing recipe code its testing the code written here. So there's no guarantee that it's testing the right thing. 

Is it possible to just call the training script under a subprocess.call() with the correct parameters / arguments? Then check the output to make sure it runs properly?

File: bionemo-recipes/recipes/codonfm_ptl_te/src/models/components/encodon_te.py (hunk 1 above)
Comment (yzhang123) [reply bionemo-recipes/recipes/codonfm_ptl_te/src/models/components/encodon_te.py:95]: we need everything on cpu and pytorch lightning moves it to gpu with trainer. FSDP otherwise doesn't work, since that will handle which layers are moved to which gpu

File: bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py (hunk 4 above)
Comment (yzhang123) [reply bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py]: ok done

File: bionemo-recipes/recipes/codonfm_ptl_te/src/utils/fsdp_config.py (hunk 2 above)
Comment (yzhang123) [reply bionemo-recipes/recipes/codonfm_ptl_te/src/utils/fsdp_config.py]: ok done

File: bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py (hunk 3 above)
Comment (yzhang123) [reply bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py]: okd one

File: bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        "src/runner.py",
+        "pretrain",
+        "--exp_name",
+        "fsdp_test",
+        "--dataset_name",
+        "SimpleCodonDataset",
+        "--process_item",
+        "mlm_memmap",
+        "--model_name",
+        "encodon_200k",
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/codonfm_ptl_te/tests/models/test_fsdp.py:57]: nit: Do we need all of these config values to be explicit or can we rely on defaults for this test?

**Comment by yzhang123** [comment]
/ok to test https://github.com/NVIDIA/bionemo-framework/commit/dd1e3207654841507afab8e478bdba4384bf1d9e
